cirq/sim/state.py

```diff
@@ -121,3 +123,157 @@ def validate_normalized_state(state: np.ndarray, num_qubits: int,
+    probs = _probs(state, indices, num_qubits)
+
+    # We now have the probability vector, correctly ordered, so sample over
+    # it. Note that we us ints here, since numpy's choice does not allow for
+    # choosing from a list of tuples or list of lists.
+    result = np.random.choice(len(probs), size=repetitions, p=probs)
+    # Convert to bools and rearrange to match repetition being the outer list.
+    return np.transpose([(1 & (result >> i)).astype(np.bool) for i in
+                         range(len(indices))]).tolist()
+
+
+def measure_state_vector(
+        state: np.ndarray,
+        indices: List[int],
+        out: np.ndarray = None) -> Tuple[List[bool], np.ndarray]:
+    """Performs a measurement of the state in the computational basis.
+
+    This does not modify `state` unless the optional `out` is `state`.
+
+    Args:
+        state: The state to be measured. This state is assumed to be normalized.
+            The state must be of size 2 ** integer.  The state can be of shape
+            (2 ** integer) or (2, 2, ..., 2).
+        indices: Which qubits are measured. The state is assumed to be supplied
+            in big endian order. That is the xth index of v, when expressed as
+            a bitstring, has the largest values in the 0th index.
+        out: An optional place to store the result. If out is the same as
+            the `state` parameter, then state will be modified inline. If out
+            is not None, then the result is put into out.  If out is None
+            a new value will be allocated. In all of these case out will be the
+            same as the returned ndarray of the method. The shape and dtype of
+            out will always match that of state.
+
+    Returns:
+        A tuple of a list and an numpy array. The list is an array of booleans
+        corresponding to the measurement values (ordered by the indices). The
+        numpy array is the post measurement state. This state has the same
+        shape and dtype as the input state.
+
+    Raises:
+        ValueError if the size of state is not a power of 2.
+        IndexError if the indices are out of range for the number of qubits
+            corresponding to the state.
+    """
+    num_qubits = _validate_num_qubits(state)
+    _validate_indices(num_qubits, indices)
+
+    if len(indices) == 0:
+        return ([], np.copy(state))
+
+    # Cache initial state.
+    initial_shape = state.shape
+
+    # Calculate the measurement probabilities and then make the measurement.
+    probs = _probs(state, indices, num_qubits)
+    result = np.random.choice(len(probs), p=probs)
+    measurement_bits = [(1 & (result >> i)) for i in range(len(indices))]
+
+    # Calculate the slice for the measurement result.
+    result_slice = linalg.slice_for_qubits_equal_to(indices, result)
+
+    # Create a mask which is False for only the slice.
+    mask = np.ones([2] * num_qubits, dtype=bool)
+    mask[result_slice] = False
```

Comment:
> That reshapes the tensor and removes dimension. I think this is faster because it doesn't have to reshape the tensor and reshuffle the memory.

I don't understand what you mean by this. How can setting a subsection of the output to zero reshape it or remove dimensions? It's an assignment of the values, not the shape, and we force the shape back to the initial shape at the end.

Reply:
```
>>> a = np.array([[1, 2], [3, 4]])
>>> a.shape
(2, 2)
>>> a[1,:]
array([3, 4])
>>> a[1,:].shape
(2,)
```
Setting a subsection to zero doesn't reshape it.  I said "that" in reference to your idea.  When you use the slice directly you end up with a new shape (see above) that excludes dimension.  If you put them back in I don't think numpy guarantees that it will fill back up the memory space correctly without shuffling things around (how could it?) Basically I'm forcing in place modification.  Remember you don't need to set the slice to zero you need to set everything but the slide to zero (hence the mask)

Reply:
> If you put them back in I don't think numpy guarantees that it will fill back up the memory space correctly without shuffling things around (how could it?)

I'm still not following what exactly you think will break.

My understanding (and the understanding that the entire `_apply_unitary_to_tensor_` code is based around) is that when you assign to a view of an array, numpy automatically takes care of scattering to the correct locations. So, for example, this code:

```python
v = np.zeros((2, 2, 2))
v[:, 1, :] = 1
```

is equivalent to this code:

```python
v = np.zeros((2, 2, 2))
v[0, 1, 0] = 1
v[0, 1, 1] = 1
v[1, 1, 0] = 1
v[1, 1, 1] = 1
```

> Remember you don't need to set the slice to zero you need to set everything but the slide to zero (hence the mask)

Ooooh, I see. Now it makes sense. Resolved.